src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexUpgradeService.java

```diff
@@ -0,0 +1,134 @@
+                    pre20HashFunction = SimpleHashFunction.class;
+                    break;
+                case "org.elasticsearch.cluster.routing.operation.hash.djb.DjbHashFunction":
+                    pre20HashFunction = DjbHashFunction.class;
+                    break;
+                default:
+                    pre20HashFunction = settings.getAsClass(DEPRECATED_SETTING_ROUTING_HASH_FUNCTION, DjbHashFunction.class, "org.elasticsearch.cluster.routing.", "HashFunction");
+            }
+        } else {
+            pre20HashFunction = DjbHashFunction.class;
+        }
+        pre20UseType = settings.getAsBoolean(DEPRECATED_SETTING_ROUTING_USE_TYPE, null);
+        if (hasCustomPre20HashFunction|| pre20UseType != null) {
+            logger.warn("Settings [{}] and [{}] are deprecated. Index settings from your old indices have been updated to record the fact that they "
+                    + "used some custom routing logic, you can now remove these settings from your `elasticsearch.yml` file", DEPRECATED_SETTING_ROUTING_HASH_FUNCTION, DEPRECATED_SETTING_ROUTING_USE_TYPE);
+        }
+    }
+
+    /**
+     * Checks that the index can be upgraded to the current version of the master node.
+     *
+     * If the index does need upgrade it returns the index metadata unchanged, otherwise it returns a modified index metadata. If index cannot be
+     * updated the method throws an exception.
+     */
+    public IndexMetaData upgradeIndexMetaData(IndexMetaData indexMetaData) throws Exception {
+        IndexMetaData newMetaData = indexMetaData;
+        newMetaData = checkSupportedVersion(newMetaData);
+        newMetaData = upgradeLegacyRoutingSettings(newMetaData);
+        return newMetaData;
+    }
+
+    /**
+     * Elasticsearch 2.0 deprecated no longer supports indices with pre Lucene v4.0 segments. All indices
+     * that were created before Elasticsearch v0.90.0 should be upgraded using upgrade plugin before they can
+     * be open by this version of elasticsearch.
+     */
+    private IndexMetaData checkSupportedVersion(IndexMetaData indexMetaData) throws Exception {
+        if (indexMetaData.getState() == IndexMetaData.State.OPEN && isSupportedVersion(indexMetaData) == false) {
+            throw new IllegalStateException("The index [" + indexMetaData.getIndex() + "] was created before v0.90.0 and wasn't upgraded."
+                    + " This index should be open using a version before " + Version.CURRENT.minimumCompatibilityVersion()
+                    + " and upgraded using upgrade API.");
+        }
+        return indexMetaData;
+    }
+
+    /*
+     * Returns true if this index can be supported by the current version of elasticsearch
+     */
+    private static boolean isSupportedVersion(IndexMetaData indexMetaData) {
+        return indexMetaData.creationVersion() != null && indexMetaData.creationVersion().onOrAfter(Version.V_0_90_0_Beta1);
```

Comment:
Maybe add minimumLuceneVersion() to index metadata as a placeholder for #11095? Then the impl can just be createdVersion().luceneVersion in this PR? It would just make it more clear here that the lucene version is the driving factor.